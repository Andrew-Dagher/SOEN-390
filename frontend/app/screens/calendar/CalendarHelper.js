// CalendarHelper.js
import * as Location from "expo-location";

/**
 * Parses location information, fetches the current device location,
 * and navigates to the appropriate screen with relevant params.
 *
 * @param {string} locationString - e.g. "SGW, Hall Building, 913"
 * @param {object} navigation - The navigation object from React Navigation.
 */
export async function handleGoToClass(locationString, navigation) {
  try {
    const [campus, buildingName] = locationString
      .split(",")
      .map((str) => str.trim());

    // Get the user's current location
    const currentLocation = await Location.getCurrentPositionAsync({});

    // Navigate to your custom "Navigation" screen (adjust as needed)
    navigation.navigate("Navigation", {
      campus: campus.toLowerCase().replace("<pre>", "").trim(),
      buildingName: buildingName
        .replace("<pre>", "")
        .replace("</pre>", "")
        .trim(),
      currentLocation: {
        latitude: currentLocation.coords.latitude,
        longitude: currentLocation.coords.longitude,
      },
    });
  } catch (error) {
    console.error("Error fetching location or parsing location string:", error);
  }
}
