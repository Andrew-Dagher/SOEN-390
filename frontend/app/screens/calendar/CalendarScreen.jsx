import React, { useState } from "react";
import { View, Text, TouchableOpacity, StyleSheet, Dimensions } from "react-native";
import { Calendar } from "react-native-calendars";
import { useNavigation } from "@react-navigation/native";
import BottomNavBar from "../../components/BottomNavBar/BottomNavBar"; // Ensure BottomNavBar is correctly imported
import CalendarDirectionsIcon from "../../components/Calendar/CalendarIcons/CalendarDirectionsIcon.jsx"; // Import CalendarDirectionsIcon

export default function CalendarScreen() {
  const navigation = useNavigation();

  // Get screen height for different devices
  const screenHeight = Dimensions.get("window").height;

  // State for the current month
  const [currentMonth, setCurrentMonth] = useState("January");

  // Function to update month based on `onMonthChange`
  const handleMonthChange = (monthData) => {
    const monthNames = [
      "January", "February", "March", "April", "May", "June",
      "July", "August", "September", "October", "November", "December"
    ];
    setCurrentMonth(monthNames[monthData.month - 1]); // Convert month number to name
  };

  return (
    <View style={{ flex: 1, backgroundColor: "#FFFFFF", flexDirection: "column" }}>
      {/* Header */}
      <View style={{ flexDirection: "row", justifyContent: "space-between", padding: 16, marginTop: "10%" }}>
        <Text style={{ fontSize: 24, color: "#000000", fontWeight: "bold" }}>{currentMonth}</Text>
        <Text style={{ fontSize: 20, color: "#E6863C" }}>27</Text>
      </View>

      {/* Calendar */}
      <Calendar
      testID="calendar-view"
        current={"2025-01-01"}
        minDate={"2025-01-01"}
        maxDate={"2025-12-31"}
        monthFormat={"yyyy MM"}
        onDayPress={(day) => console.log("selected day", day)}
        onMonthChange={handleMonthChange} // Update month on arrow press
        theme={{
          selectedDayBackgroundColor: "#E6863C",
          selectedDayTextColor: "#FFFFFF",
          todayTextColor: "#862532",
          arrowColor: "#862532",
        }}
      />

      {/* Button with Icon */}
      <View style={{ position: "absolute", bottom: "10%", left: 0, right: 0 }}>
        <TouchableOpacity
          style={styles.buttonContainer}
          onPress={() => alert("Directions are coming soon!")}
        >
          <Text style={styles.buttonText}>Get Directions to My Next Class</Text>
          <CalendarDirectionsIcon width={25} height={25} />
        </TouchableOpacity>
      </View>

      {/* Bottom Nav Bar */}
      <View style={{ position: "absolute", bottom: 0, left: 0, right: 0 }}>
        <BottomNavBar />
      </View>
    </View>
  );
}

// Button styling
const styles = StyleSheet.create({
  buttonContainer: {
    alignSelf: "center", // This will center the button horizontally
    flexDirection: "row", // Align icon and text on the same line
    alignItems: "center", // Ensure vertical alignment of text and icon
    justifyContent: "space-between",
    backgroundColor: "#862532",
    paddingVertical: 20,
    paddingHorizontal: 20,
    borderRadius: 20,
  },
  buttonText: {
    color: "#FFFFFF",
    fontWeight: "bold",
    fontSize: 16,
    marginRight: 10, // Add space between text and icon
  },
});