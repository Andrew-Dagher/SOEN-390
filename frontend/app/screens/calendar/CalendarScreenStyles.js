// CalendarScreenStyles.js
import { StyleSheet } from "react-native";

export default StyleSheet.create({
  screen: {
    flex: 1,
    backgroundColor: "#FFFFFF",
  },
  loadingContainer: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    backgroundColor: "#FFFFFF",
  },
  header: {
    flexDirection: "row",
    justifyContent: "space-between",
    padding: 16,
    marginTop: "10%",
  },
  headerText: {
    fontSize: 24,
    color: "#000000",
    fontWeight: "bold",
  },
  dateText: {
    fontSize: 20,
    color: "#E6863C",
  },
  calendarListContainer: {
    paddingHorizontal: 16,
    marginBottom: 10,
  },
  calendarButton: {
    paddingVertical: 10,
    paddingHorizontal: 15,
    backgroundColor: "#F0F0F0",
    borderRadius: 10,
    marginHorizontal: 5,
  },
  selectedCalendarButton: {
    backgroundColor: "#862532",
  },
  calendarButtonText: {
    fontSize: 16,
    color: "#000",
  },
  selectedCalendarButtonText: {
    color: "#FFFFFF",
  },
  container: {
    flex: 1,
    marginTop: 10,
    paddingHorizontal: 16,
    paddingBottom: 60,
  },
  eventBox: {
    backgroundColor: "#F5F5F5",
    padding: 20,
    marginVertical: 10,
    borderRadius: 10,
    elevation: 3,
  },
  eventTitle: {
    fontSize: 18,
    fontWeight: "bold",
    color: "#000",
  },
  eventLocation: {
    color: "#007AFF",
    marginTop: 5,
  },
  eventTime: {
    color: "#862532",
    marginTop: 5,
  },
  nextClassButton: {
    backgroundColor: "#862532",
    padding: 10,
    borderRadius: 10,
    marginTop: 10,
    alignItems: "center",
  },
  nextClassButtonText: {
    color: "#FFFFFF",
    fontWeight: "bold",
  },
  noEventsText: {
    fontSize: 16,
    color: "#555",
    textAlign: "center",
    marginTop: 20,
  },
  bottomNavBar: {
    position: "absolute",
    bottom: 0,
    left: 0,
    right: 0,
  },

  centeredDateContainer: {
    alignItems: "center",
    justifyContent: "center",
    marginVertical: 10,
  },
  centeredDateText: {
    fontSize: 20,
    fontWeight: "bold",
    color: "#E6863C",
    textAlign: "center",
  },

  dropdownContainer: {
    marginHorizontal: 16,
    marginBottom: 10,
  },
  dropdownButton: {
    backgroundColor: "#F0F0F0",
    paddingVertical: 14,
    paddingHorizontal: 16,
    borderRadius: 8,
    alignItems: "center",
    justifyContent: "center",
    width: "100%",
    shadowColor: "#000",
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.2,
    shadowRadius: 4,
    elevation: 5,
  },
  dropdownButtonText: {
    fontSize: 16,
    color: "#000",
    fontWeight: "bold",
  },
  // iOS-Style Modal Styles
  modalOverlay: {
    flex: 1,
    backgroundColor: "rgba(0, 0, 0, 0.6)",
    justifyContent: "center",
    alignItems: "center",
  },
  modalDropdown: {
    backgroundColor: "#f5f5f5",
    borderRadius: 14,
    width: "90%",
    paddingTop: 12,
    alignItems: "center",
    shadowColor: "#000",
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.3,
    shadowRadius: 6,
    elevation: 8,
    overflow: "hidden",
  },
  modalTitle: {
    fontSize: 18,
    fontWeight: "bold",
    color: "#151114",
    marginBottom: 10,
  },
  modalItem: {
    width: "100%",
    paddingVertical: 14,
    alignItems: "center",
    borderBottomWidth: 1,
    borderBottomColor: "#a1a1a1",
  },
  modalItemText: {
    fontSize: 17,
    color: "#151114",
    fontWeight: "500",
  },
  modalLastItem: {
    borderBottomWidth: 0,
  },
  modalCancel: {
    backgroundColor: "#842534",
    paddingVertical: 14,
    width: "100%",
    alignItems: "center",
    justifyContent: "center",
    borderBottomLeftRadius: 14,
    borderBottomRightRadius: 14,
  },
  modalCancelText: {
    fontSize: 16,
    fontWeight: "bold",
    color: "#FFFFFF",
  },

  guestContainer: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    backgroundColor: "#F5F5F5",
  },
  googleLoginButton: {
    flexDirection: "row",
    alignItems: "center",
    backgroundColor: "#862532",
    paddingVertical: 14,
    paddingHorizontal: 20,
    borderRadius: 12,
    elevation: 5,
    shadowColor: "#000",
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.2,
    shadowRadius: 4,
  },
  googleLoginContent: {
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "center",
    width: "100%",
  },
  googleIcon: {
    fontSize: 20,
    fontWeight: "bold",
    color: "#FFFFFF",
    backgroundColor: "#FFFFFF",
    borderRadius: 20,
    width: 30,
    height: 30,
    textAlign: "center",
    textAlignVertical: "center",
    marginRight: 10,
  },
  googleLoginText: {
    color: "#FFFFFF",
    fontSize: 16,
    fontWeight: "bold",
  },

coachMarkOverlay: {
    flex: 1,
    backgroundColor: "rgba(0, 0, 0, 0.7)", // Darker overlay for contrast
  },
  spotlight: {
    position: "absolute",
    backgroundColor: "transparent",
    borderWidth: 2,
    borderColor: "#fff",
    borderRadius: 10,
    shadowColor: "#fff",
    shadowOffset: { width: 0, height: 0 },
    shadowOpacity: 0.8,
    shadowRadius: 10,
    elevation: 5,
  },
  coachMarkContainer: {
    position: "absolute",
    top: "10%", // Adjust based on your GoToLoginButton position
    left: "10%",
    right: "10%",
    backgroundColor: "#fff",
    padding: 20,
    borderRadius: 10,
    alignItems: "center",
  },
  coachMarkContainerFinal: {
    position: "absolute",
    top: "30%",
    left: "10%",
    right: "10%",
    backgroundColor: "#fff",
    padding: 20,
    borderRadius: 10,
    alignItems: "center",
  },
  coachMarkText: {
    fontSize: 20,
    fontWeight: "bold",
    color: "#333",
    textAlign: "center",
    marginBottom: 10,
  },
  coachMarkSubText: {
    fontSize: 16,
    color: "#666",
    textAlign: "center",
    marginBottom: 20,
  },
  coachMarkButtons: {
    flexDirection: "row",
    justifyContent: "space-between",
    width: "100%",
  },
  coachMarkButton: {
    backgroundColor: "#862532",
    paddingVertical: 10,
    paddingHorizontal: 20,
    borderRadius: 5,
  },
  coachMarkButtonText: {
    color: "#fff",
    fontSize: 16,
    fontWeight: "bold",
  },
  coachMarkSkipButton: {
    paddingVertical: 10,
    paddingHorizontal: 20,
  },
  coachMarkSkipText: {
    color: "#862532",
    fontSize: 16,
    fontWeight: "bold",
  },

});
