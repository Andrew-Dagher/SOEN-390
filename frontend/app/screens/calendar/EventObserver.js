// EventObserver.js

/**
 * A simple Subject/Observer (Pub/Sub) implementation.
 * - `subscribe(fn)`: Adds an observer callback.
 * - `unsubscribe(fn)`: Removes an observer callback.
 * - `notify(data)`: Invokes all subscribed callbacks with `data`.
 */
export default class EventObserver {
  constructor() {
    // Holds an array of callback functions
    this.observers = [];
  }

  /**
   * Subscribe a callback function (observer)
   * @param {Function} fn - The callback to be invoked when notify() is called
   */
  subscribe(fn) {
    this.observers.push(fn);
  }

  /**
   * Unsubscribe a previously subscribed function
   * @param {Function} fn - The callback to remove
   */
  unsubscribe(fn) {
    this.observers = this.observers.filter((subscriber) => subscriber !== fn);
  }

  /**
   * Notify all observers, passing `data` to each callback
   * @param {*} data - The data to pass along (e.g., an array of events)
   */
  notify(data) {
    this.observers.forEach((observerFn) => observerFn(data));
  }
}
