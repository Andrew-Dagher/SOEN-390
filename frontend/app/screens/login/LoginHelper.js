import AsyncStorage from "@react-native-async-storage/async-storage";

/**
 * Fetches the calendar metadata (including the name) using the Google Calendar API.
 *
 * @param {string} calendarId - The public Google Calendar ID.
 * @returns {Promise<string>} The calendar name (summary) or a fallback name if fetch fails.
 */
export async function fetchCalendarName(calendarId, index) {
  try {
    const GOOGLE_API_KEY = process.env.EXPO_PUBLIC_GOOGLE_API_KEY2;
    if (!GOOGLE_API_KEY) {
      console.error("Google API key is missing.");
      return `Calendar ${index}`;
    }

    const url = `https://www.googleapis.com/calendar/v3/calendars/${encodeURIComponent(
      calendarId
    )}?key=${GOOGLE_API_KEY}`;

    const response = await fetch(url);
    if (!response.ok) {
      console.warn(
        `Warning: Unable to fetch calendar name for ${calendarId} (Status: ${response.status})`
      );
      return `Calendar ${index}`;
    }

    const data = await response.json();
    return data.summary || `Calendar ${index}`;
  } catch (error) {
    console.error(`Error fetching calendar name for ${calendarId}:`, error);
    return `Calendar ${index}`;
  }
}

/**
 * Fetches all available calendar IDs from environment variables and retrieves their names.
 *
 * @returns {Promise<Array<{id: string, name: string}>>} Array of calendar objects with ID and name.
 */
export async function getAvailableCalendars() {
  const calendars = [];
  let index = 1;

  // Gather all EXPO_PUBLIC_GOOGLE_CALENDAR_ID* from env
  while (true) {
    const calendarIdKey = `EXPO_PUBLIC_GOOGLE_CALENDAR_ID${index}`;
    const calendarId = process.env[calendarIdKey];
    if (!calendarId) break;

    calendars.push({
      id: calendarId,
      name: "",
    });
    index++;
  }

  // Fetch each calendar's name
  await Promise.all(
    calendars.map(async (calendar, idx) => {
      calendar.name = await fetchCalendarName(calendar.id, idx + 1);
    })
  );

  console.log("Available calendars:", calendars);
  return calendars;
}

/**
 * Fetches public Google Calendar events within a given date range.
 *
 * @param {string} calendarId - The public Google Calendar ID.
 * @param {string} startDate - Start of the range (ISO format).
 * @param {string} endDate - End of the range (ISO format).
 * @returns {Promise<Array>} Array of event objects.
 */
export async function fetchPublicCalendarEvents(calendarId, startDate, endDate) {
  try {
    const GOOGLE_API_KEY = process.env.EXPO_PUBLIC_GOOGLE_API_KEY2;
    if (!GOOGLE_API_KEY) {
      console.error("Google API key is missing.");
      return [];
    }

    const url = `https://www.googleapis.com/calendar/v3/calendars/${encodeURIComponent(
      calendarId
    )}/events?key=${GOOGLE_API_KEY}&timeMin=${startDate}&timeMax=${endDate}&singleEvents=true&orderBy=startTime`;

    console.log(`Fetching events from ${startDate} to ${endDate} for:`, calendarId);

    const response = await fetch(url);
    const textResponse = await response.text();

    if (!response.ok) {
      console.warn(
        `Warning: Unable to fetch events for ${calendarId} (Status: ${response.status})`
      );
      return [];
    }

    const data = JSON.parse(textResponse);

    if (!data.items || data.items.length === 0) {
      console.log(`No events found for ${calendarId} in the given range.`);
      return [];
    }

    return data.items.map((event, index) => ({
      id: `${calendarId}-${index + 1}`,
      title: event.summary || "No Title",
      start: {
        dateTime: event.start?.dateTime || event.start?.date || "Unknown Start",
      },
      end: {
        dateTime: event.end?.dateTime || event.end?.date || "Unknown End",
      },
      description: event.description || "No description available",
      location: event.location || "No location available",
      eventLink: event.htmlLink || "No link available",
    }));
  } catch (error) {
    console.error(`Error fetching public calendar (${calendarId}):`, error);
    return [];
  }
}

/**
 * Checks if there is an existing user session or guest mode.
 * If a session or guest mode is found, navigates to the Home screen.
 *
 * @param {Function} navigation - React Navigation function for navigation control.
 */
export async function checkExistingSession(navigation) {
  try {
    const sessionId = await AsyncStorage.getItem("sessionId");
    const guestMode = await AsyncStorage.getItem("guestMode");

    if (sessionId) {
      console.log("Existing session found, navigating to Home");
      navigation.replace("Home");
      return;
    }

    if (guestMode === "true") {
      console.log("Guest mode detected, navigating to Home");
      navigation.replace("Home");
      return;
    }
  } catch (error) {
    console.error("Error checking session:", error);
  }
}

/**
 * Stores user data (name, email, image) in AsyncStorage once the user is signed in.
 * Fetches available calendars and sets the default calendar.
 *
 * @param {Object} user - The signed-in user object.
 * @param {Function} navigation - React Navigation function for navigation control.
 */
export async function storeUserData(user, navigation) {
  if (!user) return;

  try {
    const userData = {
      fullName: user.fullName,
      email: user.primaryEmailAddress?.emailAddress,
      imageUrl: user.imageUrl,
    };

    await AsyncStorage.setItem("userData", JSON.stringify(userData));
    console.log("Stored User Data:\n", JSON.stringify(userData, null, 2));

    const calendars = await getAvailableCalendars();
    if (calendars.length > 0) {
      await AsyncStorage.setItem("availableCalendars", JSON.stringify(calendars));
      console.log("Stored available calendars:", calendars);
      await AsyncStorage.setItem("selectedCalendar", calendars[0]?.id || "");
    } else {
      console.log("No calendars found.");
    }

    navigation.replace("Home");
  } catch (error) {
    console.error("Error storing user data:", error);
  }
}

/**
 * Handles guest login by storing guest user data in AsyncStorage.
 * Navigates the user to the Home screen.
 *
 * @param {Function} navigation - React Navigation function for navigation control.
 */
export async function handleGuestLogin(navigation) {
  console.log("Guest Login Selected");
  try {
    const guestData = {
      guest: true,
      fullName: "Guest User",
      email: "guest@demo.com",
      imageUrl: null,
    };

    await AsyncStorage.setItem("guestMode", "true");
    await AsyncStorage.setItem("userData", JSON.stringify(guestData));

    navigation.replace("Home");
  } catch (error) {
    console.error("Error setting guest mode:", error);
  }
}