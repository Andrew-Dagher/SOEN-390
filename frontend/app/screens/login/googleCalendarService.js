import { useAuth } from "@clerk/clerk-expo";

export async function getGoogleAccessToken() {
  try {
    const { getToken } = useAuth(); // ‚úÖ Correct way to get the OAuth token

    if (!getToken) {
      console.error("‚ùå Clerk's getToken() method is undefined.");
      return null;
    }

    console.log("üîÑ Attempting to retrieve Google OAuth token...");

    // Attempt to retrieve OAuth token from Clerk
    const token = await getToken({ template: "oauth_google" });

    if (token) {
      console.log("‚úÖ Google Access Token Retrieved:", token);
      return token;
    } else {
      console.warn("‚ö†Ô∏è Google OAuth token is null or undefined.");
      return null;
    }
  } catch (error) {
    console.error("‚ùå Error fetching Google OAuth token:", error);
    return null;
  }
}

export async function fetchGoogleCalendarEvents() {
  const accessToken = await getGoogleAccessToken();

  if (!accessToken) {
    console.error("‚ö†Ô∏è No access token available for Google Calendar.");
    return;
  }

  try {
    const response = await fetch(
      "https://www.googleapis.com/calendar/v3/calendars/primary/events",
      {
        method: "GET",
        headers: {
          Authorization: `Bearer ${accessToken}`,
          Accept: "application/json",
        },
      }
    );

    if (!response.ok) {
      throw new Error(`‚ùå Error fetching calendar events: ${response.statusText}`);
    }

    const data = await response.json();
    console.log("‚úÖ Google Calendar Events:", data);
    return data;
  } catch (error) {
    console.error("‚ùå Error fetching Google Calendar events:", error);
    return null;
  }
}
