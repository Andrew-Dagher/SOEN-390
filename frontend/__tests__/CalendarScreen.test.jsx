import React from 'react';
import { render, fireEvent, waitFor, act } from '@testing-library/react-native';
import moment from 'moment';
import CalendarScreen from '../app/screens/calendar/CalendarScreen';
import { useAuth } from '@clerk/clerk-expo';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { fetchPublicCalendarEvents } from '../app/screens/login/LoginHelper';
import { handleGoToClass } from '../app/screens/calendar/CalendarHelper';
import { useNavigation } from '@react-navigation/native';

// Mocks for external dependencies.
jest.mock('@clerk/clerk-expo', () => ({
  useAuth: jest.fn(),
}));
jest.mock('@react-navigation/native', () => ({
  useNavigation: jest.fn(),
}));
jest.mock('@react-native-async-storage/async-storage', () => ({
  getItem: jest.fn(),
  setItem: jest.fn(),
  removeItem: jest.fn(),
}));
jest.mock('../app/screens/login/LoginHelper', () => ({
  fetchPublicCalendarEvents: jest.fn(),
}));
jest.mock('../app/screens/calendar/CalendarHelper', () => ({
  handleGoToClass: jest.fn(),
}));

// Override NotificationObserver to do nothing for these tests.
jest.mock('../app/screens/calendar/NotificationObserver', () => ({
  NotificationObserver: jest.fn(),
}));

// Use fake timers for timeout-related tests.
jest.useFakeTimers();

describe('Additional CalendarScreen tests', () => {
  let mockNavigation;
  const storedCalendars = [{ id: 'cal1', name: 'Calendar One' }];

  beforeEach(() => {
    mockNavigation = { reset: jest.fn(), navigate: jest.fn() };
    useNavigation.mockReturnValue(mockNavigation);
    jest.clearAllMocks();
  });

  // Test: When AsyncStorage returns no stored calendars.
  it('handles missing stored calendars and shows no events text', async () => {
    useAuth.mockReturnValue({ isSignedIn: true });
    // Return null for availableCalendars and selectedCalendar.
    AsyncStorage.getItem.mockResolvedValue(null);
    // fetchPublicCalendarEvents should not be called if no calendar selected.
    fetchPublicCalendarEvents.mockResolvedValue([]);

    const { getByText } = render(<CalendarScreen />);
    await waitFor(() => expect(fetchPublicCalendarEvents).not.toHaveBeenCalled());
    // Since selectedCalendar remains null, loading should be false and "No events found" displayed.
    expect(getByText('No events found for this range.')).toBeTruthy();
  });

  // Test: When loading stored calendars throws an error.
  it('logs error if loading stored calendar data fails', async () => {
    useAuth.mockReturnValue({ isSignedIn: true });
    const errorMessage = 'AsyncStorage error';
    AsyncStorage.getItem.mockRejectedValue(new Error(errorMessage));
    const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});

    render(<CalendarScreen />);
    await waitFor(() => expect(consoleErrorSpy).toHaveBeenCalledWith(
      'Error loading stored calendar data:',
      expect.any(Error)
    ));
    consoleErrorSpy.mockRestore();
  });

  // Test: Unsubscribe from observer on unmount.
  it('unsubscribes from eventsObserver on unmount', async () => {
    useAuth.mockReturnValue({ isSignedIn: true });
    // Provide stored calendars so selectedCalendar is set.
    AsyncStorage.getItem.mockImplementation((key) => {
      if (key === 'availableCalendars') return Promise.resolve(JSON.stringify(storedCalendars));
      if (key === 'selectedCalendar') return Promise.resolve('cal1');
      return Promise.resolve(null);
    });
    fetchPublicCalendarEvents.mockResolvedValue([]);

    const { unmount } = render(<CalendarScreen />);
    // Grab the eventsObserver instance from the component by spying on NotificationObserver.
    const { NotificationObserver } = require('../app/screens/calendar/NotificationObserver');
    expect(NotificationObserver).toHaveBeenCalled();
    unmount();
    // (If you had exposed unsubscribe calls on the observer instance, you could check that here.
    // In this example, we assume that unsubscribe is called in the cleanup effect.)
  });

  // Test: Toggle event expansion - pressing the same event twice collapses it.
  it('toggles event expansion when pressed twice', async () => {
    useAuth.mockReturnValue({ isSignedIn: true });
    AsyncStorage.getItem.mockImplementation((key) => {
      if (key === 'availableCalendars') return Promise.resolve(JSON.stringify(storedCalendars));
      if (key === 'selectedCalendar') return Promise.resolve('cal1');
      return Promise.resolve(null);
    });
    const event = {
      id: 'event1',
      title: 'Event 1',
      description: 'Campus A, Building 1, Room 101',
      start: { dateTime: moment().add(1, 'day').toISOString() },
      end: { dateTime: moment().add(1, 'day').add(1, 'hour').toISOString() },
    };
    fetchPublicCalendarEvents.mockResolvedValue([event]);

    const { getByText, queryByText } = render(<CalendarScreen />);
    // Wait for event to be rendered.
    await waitFor(() => expect(getByText('Event 1')).toBeTruthy());
    // Expand the event.
    fireEvent.press(getByText('Event 1'));
    expect(getByText(/📍/)).toBeTruthy();
    // Press again to collapse.
    fireEvent.press(getByText('Event 1'));
    await waitFor(() => expect(queryByText(/📍/)).toBeNull());
  });

  // Test: onRequestClose of modal hides the modal.
  it('hides the calendar selection modal when onRequestClose is triggered', async () => {
    useAuth.mockReturnValue({ isSignedIn: true });
    const calendarsData = [
      { id: 'cal1', name: 'Calendar One' },
      { id: 'cal2', name: 'Calendar Two' }
    ];
    AsyncStorage.getItem.mockImplementation((key) => {
      if (key === 'availableCalendars') return Promise.resolve(JSON.stringify(calendarsData));
      if (key === 'selectedCalendar') return Promise.resolve('cal1');
      return Promise.resolve(null);
    });
    fetchPublicCalendarEvents.mockResolvedValue([]);

    const { getByText, getByTestId, queryByText, getByA11yLabel } = render(<CalendarScreen />);
    // Open modal via the dropdown button.
    fireEvent.press(getByText('Calendar One'));
    // Simulate onRequestClose by firing the appropriate callback.
    // For this, we get the Modal component (if you assign an accessible label or testID, otherwise simulate a press on the overlay).
    // Here we simulate a press on the overlay.
    fireEvent.press(getByText('Choose a Calendar'));
    // Press the Cancel button.
    fireEvent.press(getByText('Cancel'));
    await waitFor(() => expect(queryByText('Choose a Calendar')).toBeNull());
  });

  // Test: When fetchPublicCalendarEvents returns events that all fail the regex filter.
  it('displays "No events found" if all fetched events fail the description regex', async () => {
    useAuth.mockReturnValue({ isSignedIn: true });
    AsyncStorage.getItem.mockImplementation((key) => {
      if (key === 'availableCalendars') return Promise.resolve(JSON.stringify(storedCalendars));
      if (key === 'selectedCalendar') return Promise.resolve('cal1');
      return Promise.resolve(null);
    });
    // Return events with descriptions not matching "Campus, Building, Room"
    const invalidEvents = [
      {
        id: 'event1',
        title: 'Event 1',
        description: 'Not matching format',
        start: { dateTime: moment().add(1, 'day').toISOString() },
        end: { dateTime: moment().add(1, 'day').add(1, 'hour').toISOString() },
      }
    ];
    fetchPublicCalendarEvents.mockResolvedValue(invalidEvents);

    const { getByText } = render(<CalendarScreen />);
    await waitFor(() => expect(getByText('No events found for this range.')).toBeTruthy());
  });
});