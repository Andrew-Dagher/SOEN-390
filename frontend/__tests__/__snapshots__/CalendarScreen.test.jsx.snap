// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarScreenStyles should match the snapshot 1`] = `
{
  "bottomNavBar": {
    "bottom": 0,
    "left": 0,
    "position": "absolute",
    "right": 0,
  },
  "calendarButton": {
    "backgroundColor": "#F0F0F0",
    "borderRadius": 10,
    "marginHorizontal": 5,
    "paddingHorizontal": 15,
    "paddingVertical": 10,
  },
  "calendarButtonText": {
    "color": "#000",
    "fontSize": 16,
  },
  "calendarListContainer": {
    "marginBottom": 10,
    "paddingHorizontal": 16,
  },
  "centeredDateContainer": {
    "alignItems": "center",
    "justifyContent": "center",
    "marginVertical": 10,
  },
  "centeredDateText": {
    "color": "#E6863C",
    "fontSize": 20,
    "fontWeight": "bold",
    "textAlign": "center",
  },
  "container": {
    "flex": 1,
    "marginTop": 10,
    "paddingBottom": 60,
    "paddingHorizontal": 16,
  },
  "dateText": {
    "color": "#E6863C",
    "fontSize": 20,
  },
  "dropdownButton": {
    "alignItems": "center",
    "backgroundColor": "#F0F0F0",
    "borderRadius": 8,
    "elevation": 5,
    "justifyContent": "center",
    "paddingHorizontal": 16,
    "paddingVertical": 14,
    "shadowColor": "#000",
    "shadowOffset": {
      "height": 2,
      "width": 0,
    },
    "shadowOpacity": 0.2,
    "shadowRadius": 4,
    "width": "100%",
  },
  "dropdownButtonText": {
    "color": "#000",
    "fontSize": 16,
    "fontWeight": "bold",
  },
  "dropdownContainer": {
    "marginBottom": 10,
    "marginHorizontal": 16,
  },
  "eventBox": {
    "backgroundColor": "#F5F5F5",
    "borderRadius": 10,
    "elevation": 3,
    "marginVertical": 10,
    "padding": 20,
  },
  "eventLocation": {
    "color": "#007AFF",
    "marginTop": 5,
  },
  "eventTime": {
    "color": "#862532",
    "marginTop": 5,
  },
  "eventTitle": {
    "color": "#000",
    "fontSize": 18,
    "fontWeight": "bold",
  },
  "googleIcon": {
    "backgroundColor": "#FFFFFF",
    "borderRadius": 20,
    "color": "#FFFFFF",
    "fontSize": 20,
    "fontWeight": "bold",
    "height": 30,
    "marginRight": 10,
    "textAlign": "center",
    "textAlignVertical": "center",
    "width": 30,
  },
  "googleLoginButton": {
    "alignItems": "center",
    "backgroundColor": "#862532",
    "borderRadius": 12,
    "elevation": 5,
    "flexDirection": "row",
    "paddingHorizontal": 20,
    "paddingVertical": 14,
    "shadowColor": "#000",
    "shadowOffset": {
      "height": 2,
      "width": 0,
    },
    "shadowOpacity": 0.2,
    "shadowRadius": 4,
  },
  "googleLoginContent": {
    "alignItems": "center",
    "flexDirection": "row",
    "justifyContent": "center",
    "width": "100%",
  },
  "googleLoginText": {
    "color": "#FFFFFF",
    "fontSize": 16,
    "fontWeight": "bold",
  },
  "guestContainer": {
    "alignItems": "center",
    "backgroundColor": "#F5F5F5",
    "flex": 1,
    "justifyContent": "center",
  },
  "header": {
    "flexDirection": "row",
    "justifyContent": "space-between",
    "marginTop": "10%",
    "padding": 16,
  },
  "headerText": {
    "color": "#000000",
    "fontSize": 24,
    "fontWeight": "bold",
  },
  "loadingContainer": {
    "alignItems": "center",
    "backgroundColor": "#FFFFFF",
    "flex": 1,
    "justifyContent": "center",
  },
  "modalCancel": {
    "alignItems": "center",
    "backgroundColor": "#842534",
    "borderBottomLeftRadius": 14,
    "borderBottomRightRadius": 14,
    "justifyContent": "center",
    "paddingVertical": 14,
    "width": "100%",
  },
  "modalCancelText": {
    "color": "#FFFFFF",
    "fontSize": 16,
    "fontWeight": "bold",
  },
  "modalDropdown": {
    "alignItems": "center",
    "backgroundColor": "#f5f5f5",
    "borderRadius": 14,
    "elevation": 8,
    "overflow": "hidden",
    "paddingTop": 12,
    "shadowColor": "#000",
    "shadowOffset": {
      "height": 2,
      "width": 0,
    },
    "shadowOpacity": 0.3,
    "shadowRadius": 6,
    "width": "90%",
  },
  "modalItem": {
    "alignItems": "center",
    "borderBottomColor": "#a1a1a1",
    "borderBottomWidth": 1,
    "paddingVertical": 14,
    "width": "100%",
  },
  "modalItemText": {
    "color": "#151114",
    "fontSize": 17,
    "fontWeight": "500",
  },
  "modalLastItem": {
    "borderBottomWidth": 0,
  },
  "modalOverlay": {
    "alignItems": "center",
    "backgroundColor": "rgba(0, 0, 0, 0.6)",
    "flex": 1,
    "justifyContent": "center",
  },
  "modalTitle": {
    "color": "#151114",
    "fontSize": 18,
    "fontWeight": "bold",
    "marginBottom": 10,
  },
  "nextClassButton": {
    "alignItems": "center",
    "backgroundColor": "#862532",
    "borderRadius": 10,
    "marginTop": 10,
    "padding": 10,
  },
  "nextClassButtonText": {
    "color": "#FFFFFF",
    "fontWeight": "bold",
  },
  "noEventsText": {
    "color": "#555",
    "fontSize": 16,
    "marginTop": 20,
    "textAlign": "center",
  },
  "screen": {
    "backgroundColor": "#FFFFFF",
    "flex": 1,
  },
  "selectedCalendarButton": {
    "backgroundColor": "#862532",
  },
  "selectedCalendarButtonText": {
    "color": "#FFFFFF",
  },
}
`;